pennylane/devices/default_tensor.py

```diff
@@ -889,3 +911,113 @@ def execute_and_compute_vjp(
+        device._quimb_circuit.apply_gate(
+            qml.matrix(ops).astype(device._c_dtype), *ops.wires, parametrize=None
+        )
+        return
+    theta = ops.parameters[0]
+    wire_map = dict((w, i) for i, w in enumerate(ops.wires))
+    pw = next(
+        iter(
+            qml.pauli.string_to_pauli_word(
+                ops._hyperparameters["pauli_word"], wire_map
+            ).pauli_rep.keys()
+        )
+    )
+    arrays = []
+    sites = list(pw.keys())
+    for i, P in enumerate(pw.values()):
+        if i == 0:
+            arr = qml.math.zeros((1, 2, 2, 2), dtype=complex)
+            arr[0, 0] = _PAULI_MATRICES[P]
+            arr[0, 1] = qml.math.eye(2, dtype=complex)
+
+        elif i == len(wire_map) - 1:
+            arr = qml.math.zeros((2, 1, 2, 2), dtype=complex)
+            arr[0, 0] = _PAULI_MATRICES[P] * (-1j) * qml.math.sin(theta / 2)
+            arr[1, 0] = qml.math.eye(2, dtype=complex) * qml.math.cos(theta / 2)
+
+        else:
+            arr = qml.math.zeros((2, 2, 2, 2), dtype=complex)
+            arr[0, 0] = _PAULI_MATRICES[P]
+            arr[1, 1] = qml.math.eye(2, dtype=complex)
+
+        arrays.append(arr)
+
+    mpo = qtn.MatrixProductOperator(arrays=arrays, sites=sites)
+    device._quimb_circuit._psi = mpo.apply(
+        device._quimb_circuit.psi, max_bond=device._max_bond_dim, cutoff=device._cutoff
+    )
+
+
+@apply_operation_core.register
+def apply_operation_core_trotter_product(ops: qml.TrotterProduct, device):
+    """Dispatcher for _apply_operation."""
+    time = ops.data[-1]
+    n = ops._hyperparameters["n"]
+    order = ops._hyperparameters["order"]
+    ops = ops._hyperparameters["base"].operands
+    decomp = _recursive_expression(time / n, order, ops)[::-1] * n
+    for o in decomp:
+        device._quimb_circuit.apply_gate(
+            qml.matrix(o).astype(device._c_dtype), *o.wires, parametrize=None
+        )
+
+
+@singledispatch
+def expval_core(obs: Observable, device) -> float:
+    """Dispatcher for expval."""
+    return device._local_expectation(qml.matrix(obs), tuple(obs.wires))
+
+
+@expval_core.register
+def expval_core_tensor(obs: Tensor, device) -> float:
+    """Computes the expval of a Tensor."""
+    return expval_core(Prod(*obs._args), device)
+
+
+@expval_core.register
+def expval_core_prod(obs: Prod, device) -> float:
+    """Computes the expval of a Prod."""
+    ket = copy.deepcopy(device._quimb_circuit)
+    for op in obs:
+        ket.apply_gate(qml.matrix(op).astype(device._c_dtype), *op.wires, parametrize=None)
+    return np.real((device._quimb_circuit.psi.H & ket.psi).contract(all, output_inds=()))
+
+
+@expval_core.register
+def expval_core_sprod(obs: SProd, device) -> float:
+    """Computes the expval of a SProd."""
+    return obs.scalar * expval_core(obs.base, device)
+
+
+@expval_core.register
+def expval_core_sum(obs: Sum, device) -> float:
+    """Computes the expval of a Sum."""
+    return sum(expval_core(m, device) for m in obs)
```

Comment:
It will go to either `expval_core` or `expval_core_prod`. These are the only functions that actually perform a contraction. What are you worried about?

Reply:
My worries is that the final MPS state after the circuit (`psi_final = Ufinal @ ... @ U1 psi_init`) is not yet contracted, rather a recipe for those contractions, which are then finally performed when calling `psi_final.local_expectation`. If that were the case, then in a sum of observables, `psi_final` is re-computed for every term (instead of re-using the final MPS `psi_final` that has been contracted once).

The reason I believe this is because it doesnt look like the MPS form is restored after the MPO-MPS contraction:
```python
dev = qml.device("default.tensor", wires=range(5), method="mps", max_bond_dim=100)
dev._apply_operation(qml.PauliRot(0.5, "XYZ", [0,2,4]))
psi = dev._quimb_circuit.psi

psi.tensors
```

```pycon
(Tensor(shape=(1, 1, 2), inds=('_94e9abADBAU', '_94e9abADBAV', 'k1'), tags=oset(['0', '1', '2', '3', '4', 'I1', 'PSI0'])),
 Tensor(shape=(1, 1, 2), inds=('_94e9abADBAW', '_94e9abADBAX', 'k3'), tags=oset(['0', '1', '2', '3', '4', 'I3', 'PSI0'])),
 Tensor(shape=(1, 1, 2, 2), inds=('_94e9abADBAU', '_94e9abADBAY', '_94e9abADBAZ', 'k0'), tags=oset(['0', '1', '2', '3', '4', 'I0', 'PSI0'])),
 Tensor(shape=(1, 1, 2, 2, 2), inds=('_94e9abADBAV', '_94e9abADBAW', '_94e9abADBAZ', '_94e9abADBAa', 'k2'), tags=oset(['0', '1', '2', '3', '4', 'I2', 'PSI0'])),
 Tensor(shape=(1, 2, 1, 2), inds=('_94e9abADBAX', '_94e9abADBAa', '_94e9abADBAY', 'k4'), tags=oset(['0', '1', '2', '3', '4', 'I4', 'PSI0'])))
```

it seems the virtual bonds from the MPO have not been contracted.
You can push this further:
```python
dev = qml.device("default.tensor", wires=range(5), method="mps", max_bond_dim=100)
dev._apply_operation(qml.PauliRot(0.5, "XYZ", [0,2,4]))
dev._apply_operation(qml.PauliRot(0.5, "YXZ", [1,3,4]))
dev._apply_operation(qml.PauliRot(0.5, "YXZXX", [0,1,2,3,4]))
psi = dev._quimb_circuit.psi

psi.tensors
```

```pycon
(Tensor(shape=(1, 2, 2, 2), inds=('_94e9abADBAu', '_94e9abADBAq', 'k0', '_94e9abADBAv'), tags=oset(['I0', '0', '1', '2', '3', '4', 'PSI0'])),
 Tensor(shape=(2, 2, 2, 1, 2), inds=('_94e9abADBAq', '_94e9abADBAr', 'k1', '_94e9abADBAy', '_94e9abADBAz'), tags=oset(['I1', '0', '1', '2', '3', '4', 'PSI0'])),
 Tensor(shape=(2, 2, 2, 2, 2), inds=('_94e9abADBAr', '_94e9abADBAs', 'k2', '_94e9abADBAv', '_94e9abADBAw'), tags=oset(['I2', '0', '1', '2', '3', '4', 'PSI0'])),
 Tensor(shape=(2, 4, 2, 2), inds=('_94e9abADBAs', '_94e9abADBAt', 'k3', '_94e9abADBAz'), tags=oset(['I3', '0', '1', '2', '3', '4', 'PSI0'])),
 Tensor(shape=(4, 1, 2, 1, 2), inds=('_94e9abADBAt', '_94e9abADBAu', 'k4', '_94e9abADBAy', '_94e9abADBAw'), tags=oset(['I4', '0', '1', '2', '3', '4', 'PSI0'])))
```

Perhaps this is restricted to MPO-MPS products, for regular matrix ops it seems fine

Reply:
awesome, thanks @vincentmr !